### ISSUE TITLE
[Code Change Request] codeunit 1620 "PEPPOL Validation"

### ISSUE THREAD

**Issue by KristofVervaeke** [microsoft/ALAppExtensions#21033]
In xmlport 1610 "Sales Invoice - PEPPOL BIS 3.0" lines with type = " " are excluded. (see the procedure Initialize: SalesInvoiceLine.SetFilter(Type, '<>%1', SalesInvoiceLine.Type::" ")).

However, in codeunit 1620 "PEPPOL Validation" all lines are checked, so the ones with type = " " are also checked.
So, when a customer wants to create an empty line on an invoice by creating a line with type = " " and desciption = '', then the validation codeunit tests if the description has a value. 

This is my suggestion for changing codeunit 1620.

```csharp
local procedure CheckSalesDocumentLine(SalesLine: Record "Sales Line")
    var
        GeneralLedgerSetup: Record "General Ledger Setup";
        VATPostingSetup: Record "VAT Posting Setup";
        PEPPOLMgt: Codeunit "PEPPOL Management";
        unitCode: Text;
        unitCodeListID: Text;
    begin
        if SalesLine.Type <> SalesLine.Type::" " then begin  //New Line
            PEPPOLMgt.GetLineUnitCodeInfo(SalesLine, unitCode, unitCodeListID);
            with SalesLine do begin
                if (Type <> Type::" ") and ("No." <> '') and (unitCode = '') then
                    Error(EmptyUnitOfMeasureErr, "Unit of Measure Code");
                if Description = '' then
                    Error(MissingDescriptionErr);

                if (Type <> Type::" ") and ("No." <> '') then begin // Not a description line
                    if GeneralLedgerSetup.UseVat() then
                        TestField("VAT Prod. Posting Group");
                    VATPostingSetup.Get("VAT Bus. Posting Group", "VAT Prod. Posting Group");
                    VATPostingSetup.TestField("Tax Category");
                    if (Type = Type::Item) and ("Unit Price" < 0) then
                        if not ConfirmManagement.GetResponseOrDefault(StrSubstNo(NegativeUnitPriceErr, RecordId), false) then
                            Error('');
                end;
            end;
        end;  //New Line
    end;
```

**Comment by louagej** [comment]
Can you also Provide 2 event?
1. `OnBeforeCheckSalesDocumentLine`
2. `OnAfterCheckSalesDocumentLine`

This way, we can add extra checks as well?
```al
local procedure CheckSalesDocumentLine(SalesLine: Record "Sales Line")
    var
        GeneralLedgerSetup: Record "General Ledger Setup";
        VATPostingSetup: Record "VAT Posting Setup";
        PEPPOLMgt: Codeunit "PEPPOL Management";
        unitCode: Text;
        unitCodeListID: Text;
        IsHandled: Boolean;
    begin
        OnBeforeCheckSalesDocumentLine(SalesLine, IsHandled);
        if IsHandled then
            exit;
        PEPPOLMgt.GetLineUnitCodeInfo(SalesLine, unitCode, unitCodeListID);
        if (SalesLine.Type <> SalesLine.Type::" ") and (SalesLine."No." <> '') and (unitCode = '') then
            Error(EmptyUnitOfMeasureErr, SalesLine."Unit of Measure Code");
        if SalesLine.Description = '' then
            Error(MissingDescriptionErr);

        if (SalesLine.Type <> SalesLine.Type::" ") and (SalesLine."No." <> '') then begin // Not a description line
            if GeneralLedgerSetup.UseVat() then
                SalesLine.TestField("VAT Prod. Posting Group");
            VATPostingSetup.Get(SalesLine."VAT Bus. Posting Group", SalesLine."VAT Prod. Posting Group");
            VATPostingSetup.TestField("Tax Category");
            if (SalesLine.Type = SalesLine.Type::Item) and (SalesLine."Unit Price" < 0) then
                if not ConfirmManagement.GetResponseOrDefault(StrSubstNo(NegativeUnitPriceErr, SalesLine.RecordId), false) then
                    Error('');
        end;
        OnAfterCheckSalesDocumentLine(SalesLine);
    end;
```

**Comment by b3nm4r** [comment]
Is there any update on this topic? We are having the same issue.

**Comment by mynjj** [comment]
Hi, sorry for not updating this thread appropriately after such a long time. 

I can see that the requests and events are implemented and available in versions >= 23.1.

Thanks for reporting the issue.
